src/mds/MDSRank.cc

```diff
@@ -2965,7 +2965,9 @@ void MDSRank::command_scrub_start(Formatter *f,
       recursive = true;
     else if (op == "repair")
       repair = true;
-  }
+    else if (op == "eval_stray")
+      mdcache->scan_strays();
```

Comment:
Checked the scrub code and the scrub will work as:

```
ceph tell mds.<fsname>:0 scrub start <path> [scrubopts] [tag]
```

Then you can specify the `<path>` with `~mdsdir`:

```
ceph tell mds.<fsname>:0 scrub start ~mdsdir eval_stray recursive
```

And finally in `void ScrubStack::kick_off_scrubs()` or `scrub_dir_inode()` to call the `mdcache->scan_strays()` .

If the `eval_stray` option is specified the `mdcache->scan_strays()` should only eval the stray dentries, no need to check consistency for them?

@batrick @vshankar Any better idea ?

Reply:
As discussed yesterday with @lxbsz , we can also opt for something like `ceph tell mds.<fs_name>:0 scrub eval_stray` if we don't want to interfere with scrub. (Just thinking out loud)

Reply:
Cannot recursive scrub on "/" traverse ~mdsdir?

Reply:
@dparmar18 ping?

Reply:
> Cannot recursive scrub on "/" traverse ~mdsdir?

no ... this is not possible
`/` traverses user data and `~mdsdir` traverses system data (for lack of better word) since the strays get owned by the system and don't remain accessible to the user

Reply:
> Checked the scrub code and the scrub will work as:
> 
> ```
> ceph tell mds.<fsname>:0 scrub start <path> [scrubopts] [tag]
> ```
> 
> Then you can specify the `<path>` with `~mdsdir`:
> 
> ```
> ceph tell mds.<fsname>:0 scrub start ~mdsdir eval_stray recursive
> ```
> 
> And finally in `void ScrubStack::kick_off_scrubs()` or `scrub_dir_inode()` to call the `mdcache->scan_strays()` .
> 
> If the `eval_stray` option is specified the `mdcache->scan_strays()` should only eval the stray dentries, no need to check consistency for them?
> 
> @batrick @vshankar Any better idea ?

This makes sense to me except `scrub_dirfrag` should call `maybe_eval_stray` instead if it is a scrub directory dirfrag..

Reply:
@batrick `maybe_eval_stray` would purge a dir only when that dir is marked as stray. To evaluate it we need to iterate through every item of dirfrag, mark it as stray and then call `maybe_eval_stray`. This is already being done in `MDCache::scan_stray_dir`, IMO we can just simply call it and perform evaluation.

Reply:
> > Cannot recursive scrub on "/" traverse ~mdsdir?
> 
> no ... this is not possible `/` traverses user data and `~mdsdir` traverses system data (for lack of better word) since the strays get owned by the system and don't remain accessible to the user

Sure. But my point here is to additionally traverse ~mdsdir when recursive scrub is triggered on root directory. We can additionally have the user specify ~mdsdir as <path> with `scrub start`, but that's another point. BTW, has anybody tried invoke scrub on ~mdsdir from CLI to check if MDS accepts that? (seems like it does). @mchangir @dparmar18

Reply:
@vshankar It seems MDS does accept scrub invocation from CLI (`ceph tell mds.a:0 scrub start ~mdsdir recursive`) but I'm now thinking about if it will actually evaluate strays because as @mchangir said that It might not touch system files while scrubbing.

Reply:
@dparmar18 Have you tried out the changes to see what inodes srub touches once you invoke scrub start CLI with ~mdsdir?

Reply:
scrub can iterate over ~mdsdir.